core/src/main/java/org/elasticsearch/env/Environment.java

```diff
@@ -103,30 +107,39 @@ public Environment(final Settings settings, final Path configPath) {
 
         List<String> dataPaths = PATH_DATA_SETTING.get(settings);
         final ClusterName clusterName = ClusterName.CLUSTER_NAME_SETTING.get(settings);
-        if (dataPaths.isEmpty() == false) {
-            dataFiles = new Path[dataPaths.size()];
-            dataWithClusterFiles = new Path[dataPaths.size()];
-            for (int i = 0; i < dataPaths.size(); i++) {
-                dataFiles[i] = PathUtils.get(dataPaths.get(i));
-                dataWithClusterFiles[i] = dataFiles[i].resolve(clusterName.value());
+        if (DiscoveryNode.nodeRequiresLocalStorage(settings)) {
+            if (dataPaths.isEmpty() == false) {
+                dataFiles = new Path[dataPaths.size()];
+                dataWithClusterFiles = new Path[dataPaths.size()];
+                for (int i = 0; i < dataPaths.size(); i++) {
+                    dataFiles[i] = PathUtils.get(dataPaths.get(i));
+                    dataWithClusterFiles[i] = dataFiles[i].resolve(clusterName.value());
+                }
+            } else {
+                dataFiles = new Path[]{homeFile.resolve("data")};
+                dataWithClusterFiles = new Path[]{homeFile.resolve("data").resolve(clusterName.value())};
             }
         } else {
-            dataFiles = new Path[]{homeFile.resolve("data")};
-            dataWithClusterFiles = new Path[]{homeFile.resolve("data").resolve(clusterName.value())};
+            if (dataPaths.isEmpty()) {
+                dataFiles = dataWithClusterFiles = EMPTY_PATH_ARRAY;
+            } else {
+                final String paths = String.join(",", dataPaths);
+                throw new IllegalStateException("node does not require local storage yet path.data is set to [" + paths + "]");
```

Comment:
this seems to aggressive no? if no local storage is required maybe it's just better to ignore paths altogether. In scripted deployments this could be kind of a frustrating gotcha in my experience. Maybe it's better to just keep or if it affects methods downstream maybe just set to empty and log a warning instead?

Reply:
We simply abhor allowing this kind of confusion. If a user is setting `path.data` and `node.local_storage` to false then there is a disconnect because one setting is saying store data here and the other is saying do not store data. Rather, we it to be clear: no local storage is required, so where to store data should not be set.

Reply:
> In scripted deployments this could be kind of a frustrating gotcha in my experience. 

I don't understand this comment? In a scripted environment a user would already be scripting to set `node.local_storage` to `false` (it's not the default) so why wouldn't they also script to ensure `path.data` is not set if that's what we are going to require. Fix the script once and then it's good?

Reply:
Isn't `node.local_storage` required / beneficial on client-only nodes for maintaining cluster state?

Reply:
Yes, but that does not negate either of my points.